documentation/api/collectionrules.md

```diff
@@ -0,0 +1,95 @@
+# Collection Rules - Get
+
+Reports the state of the configured collection rules of a specific process.
+
+## HTTP Route
+
+```http
+GET /collectionrules?pid={pid}&uid={uid}&name={name} HTTP/1.1
```

Comment:
I'm wondering if we should have a top level list with less details and then an api to get more information on any particular rule.

Reply:
Something like?
- `/collectionrules?pid={pid}&uid={uid}&name={name}` - returns a list of summary information for each collection for the specified process (or default process), such as an ID, name, state
- `/collectionrules/{id}` - returns detail information for a specific instance of a collection rule

The ID can be globally (within the dotnet-monitor instance) unique. So then you can use that ID to get detailed information about the rule instance. This would be easier to uniquely identify the rule as well as prevent collisions if the target process shuts down and restarts (which would have the similar process + rule name combination). This would also more easily allow "rel links" if we wanted to add those in the future.

Reply:
Any opposition to making the endpoints follow this pattern instead?

- `/collectionrules?pid={pid}&uid={uid}&name={name}`
- `/collectionrules?pid={pid}&uid={uid}&name={name}/NameOfMyRule`

This prevents users from _having_ to hit the generic endpoint to get an ID if they know what rule they want to look at, and should make the output cleaner of the generic endpoint by not having to include an extra field for a unique ID. Personally, I feel like it's a more intuitive/better user experience, but if there's other reasons why this is a bad idea, I can do IDs (or something else).

Reply:
By adding `/NameOfMyRule` to the `name` parameter, that muddles the consistent notion across all of these routes that the name parameter identifies the process name.

Reply:
While this is unlikely and strange to see, it is also possible to add any character to a file or directory name on Unix systems, such as the forward-slash. There is no "reserved" or "illegal" character notion.

Reply:
> By adding `/NameOfMyRule` to the `name` parameter, that muddles the consistent notion across all of these routes that the name parameter identifies the process name.

I don't think I'm too worried about this - perhaps being more particular about the verbiage would reduce any potential confusion, but it feels relatively intuitive (for me) to say, "for this particular process (PID/UID/Name), I want to look at the rule named xyz." I like being able to keep using the same queries over and over again (or even tab to complete them), and needing to always get the ID of a rule just seems like adding unnecessary friction by making it a two-step operation. 

In regards to your second point, I'm assuming you're saying that we could encounter something like: `/collectionrules?name=firstpartofproc/secondpartofproc/FirstPartOfCRName/SecondPartOfCRName` - you would know better than I do about whether this would be a legitimate issue, or if we could check our list of collection rule names against the back-end of the string to see if there's a match (e.g. we have a CR named "HighCpuRule/2", so we check if the last 13 characters of the string are "HighCpuRule/2").

My goal is to make this intuitive/streamlined for the 99% use-case to give the bulk of customers a better user experience, but also guaranteeing it works for that last 1% by doing the parsing (if that's an option) even if you end up with something ugly like `/collectionrules?name=firstpartofproc/secondpartofproc/FirstPartOfCRName/SecondPartOfCRName`.

Reply:
> I don't think I'm too worried about this.

I am. Consistency allows users to better predict how to use an API. If it changes slightly for arbitrary routes, that's frustrating because you have to remember "no, this route is special for some reason".

> In regards to your second point, I'm assuming you're saying that we could encounter something like: `/collectionrules?name=firstpartofproc/secondpartofproc/FirstPartOfCRName/SecondPartOfCRName` - you would know better than I do about whether this would be a legitimate issue, or if we could check our list of collection rule names against the back-end of the string to see if there's a match (e.g. we have a CR named "HighCpuRule/2", so we check if the last 13 characters of the string are "HighCpuRule/2").

Using ends-with notions leads to problems with accidental collisions and no way to disambiguate. This syntax also is hard to report errors more specific than "could not find" because we don't know where the process name ends and where the rule starts.

> My goal is to make this intuitive/streamlined for the 99% use-case to give the bulk of customers a better user experience, but also guaranteeing it works for that last 1% by doing the parsing (if that's an option) even if you end up with something ugly like `/collectionrules?name=firstpartofproc/secondpartofproc/FirstPartOfCRName/SecondPartOfCRName`.

REST-like APIs are probably the most intuitive. The operations route follows this pattern where you `GET /operations` to list all of the operations and `GET /operations/{id}` to get details of a particular operation. My proposal follows that exactly. My additional comments in my original response also enable better behaviors:
- This would be easier to uniquely identify the rule as well as prevent collisions if the target process shuts down and restarts (which would have the similar process + rule name combination).
- This would also more easily allow "rel links" if we wanted to add those in the future.

Additionally, we cannot represent two different structured data formats for the same verb (GET) and HTTP response code (200) pairing in the openapi document. So we cannot represent a JSON object with collection rule names as the properties and another JSON object with the detail fields as its properties for the `GET /collectionrules -> 200` pairing.

Reply:
Alternatively, something like `GET /collectionrules/{name}?pid={pid}&uid={uid}&name={name}` would allow you get the collection rule instance for the matching process or the default process if `GET /collectionrules/{name}` as issued.

Reply:
> Alternatively, something like `GET /collectionrules/{name}?pid={pid}&uid={uid}&name={name}` would allow you get the collection rule instance for the matching process or the default process if `GET /collectionrules/{name}` is issued.

I had considered this, but wasn't sure if that would be more jarring by not having the process determination first. Purely for the purposes of not having to repeatedly look up a changing ID, I prefer this though over the initial suggestion - but I'll ultimately defer to what you think makes the most sense for the API, I just wanted to voice my concerns.